### PR TITLE
Parler tts release

### PR DESCRIPTION
**Pull request by ylacombe**
Prepare release

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,93 +1,104 @@
...
-```
+# Parler-TTS
+
+[[Paper we reproduce]](https://arxiv.org/abs/2402.01912)
+[[Models]](https://huggingface.co/parler-tts)
+[[Training Code]](training)
+[[Interactive Demo]](https://huggingface.co/spaces/parler-tts/parler_tts_mini)
+
+> [!IMPORTANT]
+> We're proud to release Parler-TTS v0.1, our first 300M-parameters Parler-TTS model, trained on 10.5K hours of audio data.
```
Comment (sanchit-gandhi) [inline README.md:9]:
```suggestion
> We're proud to release Parler-TTS v0.1, our first 300M parameter model, trained on 10.5K hours of audio data.
> In the coming weeks, we'll be working on scaling up to 50k hours of data, in preparation for the v1 model.
```

File: README.md [hunk 2]
```diff
@@ -1,93 +1,104 @@
...
+
+Parler-TTS is a reproduction of the text-to-speech (TTS) model from the paper [Natural language guidance of high-fidelity text-to-speech with synthetic annotations](https://www.text-description-to-speech.com)
+by Dan Lyth and Simon King, from Stability AI and Edinburgh University respectively. 
+
+Contrarily to standard TTS models, Parler-TTS allows you to directly describe the speaker characteristics with a simple text description where you can modulate gender, pitch, speaking style, accent, etc.
```
Comment (sanchit-gandhi) [inline README.md:14]: Maybe what's more contrary to standard TTS models is the fact that **everything** about this release is open-source? The datasets, the pre-processing, the training code and weights => this is the first truly open-source release of a TTS model we've had in a long time. IMO you can make this a selling point of this work to promote the community to use the tools as much as they do the models

```suggestion
Contrarily to other TTS models, Parler-TTS is a **fully open-source** release. All of the datasets, pre-processing, training code and weights are released publicly under permissive license, enabling the community to build on our work and develop their own powerful TTS models.

This repository contains the inference and training code for Parler-TTS. It is designed to accompany the [Data-Speech](https://github.com/ylacombe/dataspeech) repository for dataset annotation.
```

File: README.md [hunk 3]
```diff
@@ -1,93 +1,104 @@
...
+sf.write("parler_tts_out.wav", audio_arr, model.config.sampling_rate)
+```
 
-## Step 3: Aggregate Statistics
 
-Aggregate statistics from Step 2. Convert continuous values to discrete labels.
+## Installation steps
 
-## Step 4: Create Descriptions
+Parler-TTS has light-weight dependencies and can be installed in one line:
```
Comment (sanchit-gandhi) [inline README.md:45]: Are we set on doing this as a PyPi package? Or just installation through GitHub?

File: README.md [hunk 4]
```diff
@@ -1,93 +1,104 @@
...
+```sh
+pip install parler-tts
+```
 
-Convert sequence of discrete labels to text description (using an LLM). 
+## Gradio demo
```
Comment (sanchit-gandhi) [inline README.md:50]: IMO this is less interesting than the training steps -> the Gradio demo can be done with just the tab at the top of the README, since it's not really the focus of the release.

Instead, I would replace this with a section on `Training`. You can then either put the training steps in the main README here, or add a link to the subfolder on training if you think it will get too long. I would have a preference for the former, since we're actively trying to encourage people to use our training code.

File: README.md [hunk 5]
```diff
@@ -1,93 +1,104 @@
...
+```
+
+## Acknowledgements
+
+This library builds on top of a number of open-source giants, to whom we'd like to extend our warmest thanks for providing these tools!
+
+Special thanks to:
+- Dan Lyth and Simon King, from Stability AI and Edinburgh University respectively, for publishing such a promising and clear research paper: [Natural language guidance of high-fidelity text-to-speech with synthetic annotations](https://arxiv.org/abs/2402.01912).
+- the many libraries used, namely [🤗 datasets](https://huggingface.co/docs/datasets/v2.17.0/en/index), [🤗 accelerate](https://huggingface.co/docs/accelerate/en/index), [jiwer](https://github.com/jitsi/jiwer), [wandb](https://wandb.ai/), and [🤗 transformers](https://huggingface.co/docs/transformers/index).
+- HuggingFace 🤗 for providing compute resources and time to explore!
```
Comment (sanchit-gandhi) [inline README.md:71]:
```suggestion
- Descript for the [DAC codec model](https://github.com/descriptinc/descript-audio-codec)
- Hugging Face 🤗 for providing compute resources and time to explore!
```

File: README.md [hunk 6]
```diff
@@ -1,93 +1,104 @@
...
+    - Explore multilingual training.
+    - Explore mono-speaker finetuning.
+    - Explore more architectures.
+- Optimization:
+    - Compilation and static cache
+    - Support to FA2 and SDPA
+- Evaluation:
+    - Add more evaluation metrics
+
+## Citation
```
Comment (sanchit-gandhi) [inline README.md:94]:
```suggestion
## Citation
If you found this repository useful, please consider citing this work and also the original Stability AI paper:
```

File: README.md [hunk 7]
```diff
@@ -1,93 +1,104 @@
...
+```
+@misc{lacombe-etal-2024-parler-tts,
+  author = {Yoach Lacombe and Vaibhav Srivastav and Sanchit Gandhi},
+  title = {Parler-TTS},
+  year = {2024},
+  publisher = {GitHub},
+  journal = {GitHub repository},
+  howpublished = {\url{https://github.com/huggingface/parler-tts}}
+}
+```
```
Comment (sanchit-gandhi) [inline README.md:104]: Let's also add a citation for the stability paper here since we leverage their work very heavily (I can't do this as a GH suggestion due to the triple backticks that are required)

```
@misc{lyth2024natural,
      title={Natural language guidance of high-fidelity text-to-speech with synthetic annotations}, 
      author={Dan Lyth and Simon King},
      year={2024},
      eprint={2402.01912},
      archivePrefix={arXiv},
      primaryClass={cs.SD}
}
```

File: README.md [hunk 8]
```diff
@@ -1,93 +1,104 @@
...
+## Usage
+
+> [!TIP]
+> You can directly try it out in an interactive demo [here](https://huggingface.co/spaces/parler-tts/parler_tts_mini)!
+
+Using Parler-TTS is as simple as "bonjour". Simply use the following inference snippet.
+
+```py
+from parler_tts import ParlerTTSForConditionalGeneration
+from transformers import AutoTokenizer, AutoFeatureExtractor
```
Comment (sanchit-gandhi) [inline README.md:25]:
```suggestion
from transformers import AutoTokenizer
```

File: README.md [hunk 9]
```diff
@@ -1,93 +1,104 @@
...
 
-## Step 2: Annotate the Training Set
+prompt = "Hey, how are you doing today?"
+description = "A female speaker with a slightly low-pitched voice delivers her words quite expressively, in a very confined sounding environment with clear audio quality. She speaks very fast."
 
-Annotate the training dataset with information on: SNR, C50, pitch and speaking rate. 
+input_ids = tokenizer(description, return_tensors="pt").input_ids
+prompt_input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+generation = model.generate(input_ids=input_ids, prompt_input_ids=prompt_input_ids)
```
Comment (sanchit-gandhi) [inline README.md:37]: Nice that you set `do_sample=True` in the generation config already!

Do you think this example should also select the GPU if available?

File: training/TRAINING.md [hunk 10]
```diff
@@ -0,0 +1,213 @@
+# Training Parler-TTS
+
+This sub-folder contains all the information to train or finetune you own Parler-TTS model. It consists in:
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:3]:
```suggestion
This sub-folder contains all the information to train or fine-tune your own Parler-TTS model. It consists of:
```

File: training/TRAINING.md [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+- [A. An introduction to Parler-TTS architecture](#a-architecture)
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:4]:
```suggestion
- [A. An introduction to the Parler-TTS architecture](#a-architecture)
```

File: training/TRAINING.md [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+- [B. First steps to get started](#b-getting-started)
+- [C. Training guide](#c-training)
+- [E. Scaling up to 10.5K hours](#d-scaling-up---discussions-and-tips)
+
+
+## A. Architecture
+
+At the moment, Parler-TTS architecture is a carbon copy of [Musicgen architecture](https://huggingface.co/docs/transformers/v4.39.3/en/model_doc/musicgen#model-structure) and can be decomposed into three distinct stages:
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:12]:
```suggestion
At the moment, Parler-TTS architecture is a carbon copy of the [MusicGen architecture](https://huggingface.co/docs/transformers/v4.39.3/en/model_doc/musicgen#model-structure) and can be decomposed into three distinct stages:
```

File: training/TRAINING.md [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+>1. Text encoder: maps the text inputs to a sequence of hidden-state representations. The pre-trained MusicGen models use a frozen text encoder from either T5 or Flan-T5
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:13]:
```suggestion
>1. Text encoder: maps the text descriptions to a sequence of hidden-state representations. Parler-TTS uses a frozen text encoder initialised entirely from Flan-T5
```

File: training/TRAINING.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+>2. Parler-TTS decoder: a language model (LM) that auto-regressively generates audio tokens (or codes) conditional on the encoder hidden-state representations
+>3. Audio encoder: used to recover the audio waveform from the audio tokens predicted by the decoder
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:15]:
```suggestion
>3. Audio codec: used to recover the audio waveform from the audio tokens predicted by the decoder. We use the [DAC model](https://github.com/descriptinc/descript-audio-codec) from Descript, although other codec models, such as [EnCodec](https://huggingface.co/facebook/encodec_48khz), can also be used
```

File: training/TRAINING.md [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+The Parler-TTS code is written in [PyTorch](https://pytorch.org) and [Accelerate](https://huggingface.co/docs/accelerate/index). It uses some additional requirements, like [wandb](https://wandb.ai/), especially for logging and evaluation.
+
+To install the package for training, you need to clone the repository from source...
+
+```bash
+git clone https://github.com/huggingface/parler-tts.git
+cd parler-tts
+```
+
+... And then to install requirements.
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:41]:
```suggestion
... And then install the requirements:
```

File: training/TRAINING.md [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+```bash
+pip install -e .[train]
+```
+
+Optionnally, you can create a wandb account and login to it by following [this guide](https://docs.wandb.ai/quickstart). [`wandb`](https://docs.wandb.ai/) allows for better tracking of the experiments metrics and losses.
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:47]:
```suggestion
Optionally, you can create a wandb account and login to it by following [this guide](https://docs.wandb.ai/quickstart). [`wandb`](https://docs.wandb.ai/) allows for better tracking of the experiments metrics and losses.
```

File: training/TRAINING.md [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+    --dataloader_num_workers 8
+```
+
+
+> [!TIP]
+> Fine-tuning is as easy as modifying `model_name_or_path` to a pre-trained model.
+> For example: `--model_name_or_path parler-tts/parler_tts_300M_v0.1`.
+
+
+Additionnally, you can also write a JSON config file. Here, [librispeech_tts_r_300M_dummy.json](/helpers/training_configs/librispeech_tts_r_300M_dummy.json) contains the exact same hyper-parameters than above and can be launched like that:
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:167]:
```suggestion
Additionally, you can also write a JSON config file. Here, [librispeech_tts_r_300M_dummy.json](/helpers/training_configs/librispeech_tts_r_300M_dummy.json) contains the exact same hyper-parameters than above and can be launched like that:
```

File: training/TRAINING.md [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+    --target_audio_column_name "audio" \
+    --description_column_name "text_description" \
+    --prompt_column_name "text" \
+    --max_duration_in_seconds 20 \
+    --min_duration_in_seconds 2.0 \
+    --add_audio_samples_to_wandb true \
+    --id_column_name "id" \
+    --preprocessing_num_workers 8 \
+    --pad_token_id 1024 \
+    --decoder_start_token_id 1025 \
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:128]: Decoder start and pad token ids could be set in the model init script right? They seem more relevant there no?

File: training/TRAINING.md [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+
+## C. Training
+
+The script [`run_parler_tts_training.py`](/training/run_parler_tts_training.py) is an end-to-end script that:
+1. load dataset(s) and merge them to the annotation dataset(s) if necessary
+2. pre-compute audio tokens
+3. train Parler-TTS
+
+In this example, we will train and evaluate on a subsample of the test split. This is purely to demonstrate how to use the training script, rather than recommended advice for defining train/validation splits. We advise that you train on the train splits of your dataset, evaluate and tune hyper-parameters on the validation split, and only test the final checkpoint on the test split.
```
Comment (sanchit-gandhi) [inline training/TRAINING.md:101]: From experience it's better to provide an example that actually gets reasonable results. I've tried in the past giving lower resource dummy examples, and what ends up happening is that some users can't figure out how to adapt them for actual training runs -> my advice here would be to include the args that will give proper training results, ideally those that were used to train the v0.1 model. This way, people can build on your work most easily

File: training/run_parler_tts_training.py [hunk 20]
```diff
@@ -0,0 +1,1759 @@
+#!/usr/bin/env python
```
Comment (sanchit-gandhi) [inline training/run_parler_tts_training.py:1]: This looks really good!

File: parler_tts/modeling_parler_tts.py [hunk 21]
```diff
@@ -1041,92 +1218,65 @@ def prepare_inputs_for_generation(
...
+            "prompt_attention_mask": prompt_attention_mask,
             "head_mask": head_mask,
             "cross_attn_head_mask": cross_attn_head_mask,
             "past_key_values": past_key_values,
             "use_cache": use_cache,
         }
 
     # Ignore copy
-    def build_delay_pattern_mask(self, input_ids: torch.LongTensor, pad_token_id: int, max_length: int = None):
+    def build_delay_pattern_mask(
```
Comment (sanchit-gandhi) [inline parler_tts/modeling_parler_tts.py:1251]: Is there a reason we essentially make this a function instead of method?

File: parler_tts/modeling_parler_tts.py [hunk 22]
```diff
@@ -1536,15 +1704,15 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
         Example:
 
         ```python
-        >>> from transformers import StableSpeechForConditionalGeneration
+        >>> from transformers import ParlerTTSForConditionalGeneration
```
Comment (sanchit-gandhi) [inline parler_tts/modeling_parler_tts.py:1707]:
```suggestion
        >>> from parler_tts import ParlerTTSForConditionalGeneration
```

File: parler_tts/modeling_parler_tts.py [hunk 23]
```diff
@@ -1615,18 +1783,18 @@ def from_sub_models_pretrained(
...
```
Comment (sanchit-gandhi) [inline parler_tts/modeling_parler_tts.py:1786]:
```suggestion
        >>> from parler_tts import ParlerTTSForConditionalGeneration
```

File: parler_tts/modeling_parler_tts.py [hunk 24]
```diff
@@ -2430,59 +2633,54 @@ def generate(
...
+            for sample_id in range(batch_size):
+                sample = output_ids[:, sample_id]
+                sample_mask = (sample >= self.audio_encoder.config.codebook_size).sum(dim=(0, 1)) == 0
+                if sample_mask.sum() > 0:
+                    sample = sample[:, :, sample_mask]
+                    sample = self.audio_encoder.decode(sample[None, ...], [audio_scales[sample_id]]).audio_values
+                    output_values.append(sample.transpose(0, 2))
+                else:
+                    output_values.append(torch.zeros((1, 1, 1)).to(self.device))
+            # TODO: we should keep track of output length as well. Not really straightfoward tbh
```
Comment (sanchit-gandhi) [inline parler_tts/modeling_parler_tts.py:2675]: Return a tensor of shape `(bsz,)` that contains the output lengths as integers?

**Review by sanchit-gandhi — APPROVED** [review]
Well done on getting the repo cleaned up, it's looking in good shape.

My opinion is that the main README should bring visibility to both inference and training, since we want to push this as a toolkit to train new TTS models. In that regard, it would be great to include the training steps in the main REAMDE, or at least a reference to the training README if too long.

For the training README, let's also try to keep it quite streamlined: one command to run per-section, with the args set to reproduce the v0.1 checkpoint. This way, it'll be easier for the community to build on your work.

Leaving this as another TODO here: revisit the generation code (forcing ids and ensuring we don't get early stopping)

File: parler_tts/modeling_parler_tts.py (hunk 21 above)
Comment (ylacombe) [reply parler_tts/modeling_parler_tts.py:1251]: Yes indeed, I'm using it in a `Dataset.map` and this avoid hashing issues!

File: parler_tts/modeling_parler_tts.py (hunk 24 above)
Comment (ylacombe) [reply parler_tts/modeling_parler_tts.py:2675]: This was more about passing batch to DAC instead of sequential decoding!

File: parler_tts/modeling_parler_tts.py (hunk 21 above)
Comment (sanchit-gandhi) [reply parler_tts/modeling_parler_tts.py:1251]: I think you can do something hacky in your training script where you do:

```python
build_delay_pattern_mask = model.build_delay_pattern_mask

def preprocess(batch):
    # do some preprocessing with the pattern mask
    batch["labels"] = build_delay_pattern_mask(labels)
    return batch

dataset = dataset.map(preprocess)
```

But this is also a sufficient workaround as well

File: parler_tts/modeling_parler_tts.py (hunk 24 above)
Comment (sanchit-gandhi) [reply parler_tts/modeling_parler_tts.py:2675]: Ah I see - yeah passing a mask is quite involved, let's leave this as a TODO. Having the LM work with SDPA/FA2 is going to bring far greater inference speed gains than batched decoding with DAC

File: README.md (hunk 3 above)
Comment (ylacombe) [reply README.md:48]: Installation though GitHub following our private discussion

File: README.md (hunk 9 above)
Comment (ylacombe) [reply README.md:40]: I was thinking about it too! Done !
